File: day13.py
"""Solution to day 13 of Advent of Code"""

from get_input import get_input
import math
import itertools


def part1(lines):
    start, buses = lines
    next_arrival = (None, float("inf"))
    for bus in buses:
        if bus == 'x':
            continue
        next_arrival = min(
            next_arrival,
            (bus, math.ceil(start / bus) * bus),
            key=lambda n: n[1]
        )
    return next_arrival[0] * (next_arrival[1] - start)


def bus_lcm(bus_a, bus_b):
    period = bus_a[1] * bus_b[1] // math.gcd(bus_a[1], bus_b[1])
    for offset in itertools.count(bus_a[0], bus_a[1]):
        if (offset - bus_b[0]) % bus_b[1] == 0:
            return (offset, period)


def part2(lines):
    buses = list(e for e in enumerate(lines[1]) if e[1] != "x")
    final = buses[0]
    for bus in buses[1:]:
        final = bus_lcm(final, bus)
    return final[1] - final[0]


def parse(text):
    lines = text.splitlines()
    start = int(lines[0].strip())
    buses = [
        int(val) if val != 'x' else val
        for val in lines[1].strip().split(',')
    ]
    return (start, buses)


TEST = """939
7,13,x,x,59,x,31,19"""

if __name__ == "__main__":
    assert part1(parse(TEST)) == 295
    assert part2((None, [17, 'x', 13, 19])) == 3417
    assert part2(parse(TEST)) == 1068781
    LINES = parse(get_input(day=13, year=2020))
    print(f"Part 1: {part1(LINES)}")
    print(f"Part 2: {part2(LINES)}")

File: day13.rs
use common::get_input;
use num::BigInt;
use std::cmp;

type Input = (usize, Vec<Option<usize>>);

fn part1(input: &Input) -> isize {
    let start = input.0.clone();
    let next = input.1.iter()
        .filter_map(|&b| b)
        .map(|b| {
            (b, b * (start/ b + 1))
        })
        .min_by_key(|&b| b.1)
        .unwrap();
    (next.0 * (next.1 - start)) as isize
}

type Bus = (BigInt, usize);

fn bus_lcm(bus_a: &Bus, bus_b: &Bus) -> Bus {
    let mut a = cmp::max(bus_a.1, bus_b.1);
    let mut b = cmp::min(bus_a.1, bus_b.1);
    while b > 0 {
        let t = b;
        b = a % b;
        a = t;
    }
    let period = (bus_a.1 * bus_b.1) / a;
    let mut offset: BigInt = bus_a.0.clone();
    loop {
        offset += bus_a.1;
        if (offset.clone() - bus_b.0.clone()) % BigInt::from(bus_b.1) == BigInt::from(0) {
            return (offset, period)
        }
    }
}

fn part2(buses: &Input) -> BigInt {
    let mut bus_list = buses.1.iter().enumerate()
        .filter_map(|bus| if let (b, Some(num)) = bus { Some((BigInt::from(b), *num)) } else { None });
    let mut last = bus_list.next().map(|b| (BigInt::from(b.0), b.1)).unwrap();
    for bus in bus_list {
        last = bus_lcm(&last, &bus);
    }
    last.1 - last.0
}

fn parse(text: String) -> Input {
    let mut lines = text.trim().split('\n').map(|l| l.trim());
    (
        lines.next().unwrap().parse::<usize>().unwrap(),
        lines.next().unwrap().split(',')
            .map(|line| if line == "x" { None } else { Some(line.parse::<usize>().unwrap()) })
            .collect()
    )
}

fn main() {
    let input = parse(get_input(13, 2020));
    println!("Part 1: {}", part1(&input));
    println!("Part 2: {}", part2(&input));

}
